{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# include path to RoboCOP directory\n",
    "import sys\n",
    "sys.path.insert(0, '../pkg/')\n",
    "import matplotlib.pyplot as plt\n",
    "from robocop_diff import nuc_diff_map, tf_diff_map, get_diff_tfs, annotate_nucs\n",
    "from robocop_diff.robocop_diff_plot import plot_diff_cop"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We first ran RoboCOP on 5 separate MNase-seq data sets of cadmium treatment. Here are the names of the five RoboCOP output directories."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "dirname1 = '../../../robocop_cd_DM504_Chr1_16/'\n",
    "dirname2 = '../../../robocop_cd_DM505_Chr1_16_new/'\n",
    "dirname3 = '../../../robocop_cd_DM506_Chr1_16_new/'\n",
    "dirname4 = '../../../robocop_cd_DM507_Chr1_16/'\n",
    "dirname5 = '../../../robocop_cd_DM508_Chr1_16/'\n",
    "dirnames = [dirname1, dirname2, dirname3, dirname4, dirname5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# output directory for DynaCOP output\n",
    "outdir = '../../../robocop_diff_cd_DM504_DM505_DM506_DM507_DM508/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create nucleosome linkage map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create nucleosome linkage map by calling ```nuc_map_multiple```. The function first calls nucleosome dyads individually from the RoboCOP outputs. It then links the nucleosomes across the outputs to form a genome-wide nucleosome map for the given sets of data. Note that the nucleosomes are linked with an assumption that there exists an order in the set out RoboCOP outputs. For example, in our case we have a time series experiment of cadmium treatment of yeast cells. In case there is no order, the dirnames need to be provided in a sequence that the users seem fit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "nuc_df = nuc_diff_map.nuc_map_multiple(dirnames, outdir)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns of ```nuc_df```:\n",
    "\n",
    "* **chr:** chromosome\n",
    "\n",
    "* **dyadX:** nucleosome dyad; the dyad position in different RoboCOP outputs are denoted as dyadA, dyadB,...\n",
    "\n",
    "* **occupancyX:** Nucleosome posterior as predicted by RoboCOP; the occupancy in the different data sets are denoted as occupancyA, occupancyB,...\n",
    "\n",
    "* **pdyadX:** Nucleosome dyad posterior as predicted by RoboCOP; the dyaad posterior is denoted as pdyadA, pdyadB, ...\n",
    "\n",
    "* **shift_XY:** Shift in nucleosome dyad between two consecutive RoboCOP outputs. Shift calculated as dyadY - dyadX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>dyadA</th>\n",
       "      <th>dyadB</th>\n",
       "      <th>occupancyA</th>\n",
       "      <th>occupancyB</th>\n",
       "      <th>pdyadA</th>\n",
       "      <th>pdyadB</th>\n",
       "      <th>shift_AB</th>\n",
       "      <th>dyadC</th>\n",
       "      <th>occupancyC</th>\n",
       "      <th>...</th>\n",
       "      <th>pdyadC</th>\n",
       "      <th>dyadD</th>\n",
       "      <th>occupancyD</th>\n",
       "      <th>shift_CD</th>\n",
       "      <th>pdyadD</th>\n",
       "      <th>dyadE</th>\n",
       "      <th>occupancyE</th>\n",
       "      <th>shift_DE</th>\n",
       "      <th>pdyadE</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>229.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>0.999942</td>\n",
       "      <td>0.984759</td>\n",
       "      <td>0.315144</td>\n",
       "      <td>0.182249</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>0.999783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.334864</td>\n",
       "      <td>222.0</td>\n",
       "      <td>0.975525</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>0.163414</td>\n",
       "      <td>229.0</td>\n",
       "      <td>0.999758</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.251794</td>\n",
       "      <td>nuc_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>389.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.999969</td>\n",
       "      <td>0.996902</td>\n",
       "      <td>0.270605</td>\n",
       "      <td>0.136205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.997403</td>\n",
       "      <td>...</td>\n",
       "      <td>0.170422</td>\n",
       "      <td>385.0</td>\n",
       "      <td>0.997682</td>\n",
       "      <td>-19.0</td>\n",
       "      <td>0.277604</td>\n",
       "      <td>381.0</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>0.275946</td>\n",
       "      <td>nuc_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>866.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>0.999912</td>\n",
       "      <td>0.427427</td>\n",
       "      <td>0.518984</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.636082</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.462330</td>\n",
       "      <td>865.0</td>\n",
       "      <td>0.999345</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.277971</td>\n",
       "      <td>nuc_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.889191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuc_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1582.0</td>\n",
       "      <td>1588.0</td>\n",
       "      <td>0.999886</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.310744</td>\n",
       "      <td>0.247687</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1580.0</td>\n",
       "      <td>0.998400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.277184</td>\n",
       "      <td>1582.0</td>\n",
       "      <td>0.999767</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.149365</td>\n",
       "      <td>1580.0</td>\n",
       "      <td>0.997392</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.153465</td>\n",
       "      <td>nuc_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   dyadA   dyadB  occupancyA  occupancyB    pdyadA    pdyadB  shift_AB  \\\n",
       "0  chrI   229.0   216.0    0.999942    0.984759  0.315144  0.182249     -13.0   \n",
       "1  chrI   389.0   389.0    0.999969    0.996902  0.270605  0.136205       0.0   \n",
       "2  chrI   866.0   864.0    0.999993    0.999912  0.427427  0.518984      -2.0   \n",
       "3  chrI  1179.0     NaN    0.889191         NaN  0.272899       NaN       NaN   \n",
       "4  chrI  1582.0  1588.0    0.999886    0.999997  0.310744  0.247687       6.0   \n",
       "\n",
       "    dyadC  occupancyC  ...    pdyadC   dyadD  occupancyD  shift_CD    pdyadD  \\\n",
       "0   230.0    0.999783  ...  0.334864   222.0    0.975525      -8.0  0.163414   \n",
       "1   404.0    0.997403  ...  0.170422   385.0    0.997682     -19.0  0.277604   \n",
       "2   864.0    0.999965  ...  0.636082   864.0    0.999907       0.0  0.462330   \n",
       "3     NaN         NaN  ...       NaN     NaN         NaN       NaN       NaN   \n",
       "4  1580.0    0.998400  ...  0.277184  1582.0    0.999767       2.0  0.149365   \n",
       "\n",
       "    dyadE  occupancyE  shift_DE    pdyadE   name  \n",
       "0   229.0    0.999758       7.0  0.251794  nuc_0  \n",
       "1   381.0    0.999992      -4.0  0.275946  nuc_1  \n",
       "2   865.0    0.999345       1.0  0.277971  nuc_2  \n",
       "3     NaN         NaN       NaN       NaN  nuc_3  \n",
       "4  1580.0    0.997392      -2.0  0.153465  nuc_4  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuc_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create TF difference table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get TF binding sites that bind with probability > 0.1 in at least one of the RoboCOP outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_sites = get_diff_tfs.get_tf_diff(dirnames, outdir) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns of ```tf_sites```:\n",
    "    \n",
    "* **chr:** chromosome\n",
    "        \n",
    "* **start:** start of TF binding site\n",
    "        \n",
    "* **end**: end of TF binding site\n",
    "        \n",
    "* **scoreX:** posterior of TF binding at predicted by RoboCOP\n",
    "        \n",
    "* **TF:** name of TF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>scoreA</th>\n",
       "      <th>scoreB</th>\n",
       "      <th>scoreC</th>\n",
       "      <th>scoreD</th>\n",
       "      <th>scoreE</th>\n",
       "      <th>TF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>460398.0</td>\n",
       "      <td>460405.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>YRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>460370.0</td>\n",
       "      <td>460377.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>YRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>459967.0</td>\n",
       "      <td>459974.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>YRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>459860.0</td>\n",
       "      <td>459867.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>YRM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrXII</td>\n",
       "      <td>460384.0</td>\n",
       "      <td>460391.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>YRM1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chr     start       end  scoreA  scoreB  scoreC  scoreD  scoreE    TF\n",
       "0  chrXII  460398.0  460405.0     1.0     0.0     0.0     0.0     0.0  YRM1\n",
       "1  chrXII  460370.0  460377.0     1.0     0.0     0.0     0.0     0.0  YRM1\n",
       "2  chrXII  459967.0  459974.0     1.0     0.0     0.0     0.0     0.0  YRM1\n",
       "3  chrXII  459860.0  459867.0     1.0     0.0     0.0     0.0     0.0  YRM1\n",
       "4  chrXII  460384.0  460391.0     1.0     0.0     0.0     0.0     0.0  YRM1"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_sites.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plot DynaCOP output for a segment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "*c* argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with *x* & *y*.  Please use the *color* keyword-argument or provide a 2-D array with a single row if you intend to specify the same RGB or RGBA value for all points.\n",
      "../pkg/robocop_diff/robocop_diff_plot.py:109: RuntimeWarning: All-NaN axis encountered\n",
      "  nuc_df['min_dyad'] = np.nanmin(nuc_df[dyads], axis=1)\n",
      "../pkg/robocop_diff/robocop_diff_plot.py:110: RuntimeWarning: All-NaN axis encountered\n",
      "  nuc_df['max_dyad'] = np.nanmax(nuc_df[dyads], axis=1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Output saved: ../../../robocop_diff_cd_DM504_DM505_DM506_DM507_DM508/figures/MET10.pdf\n"
     ]
    }
   ],
   "source": [
    "chrm = 'chrVI'\n",
    "start = 212000                                                                                                                                                                         \n",
    "end = 213800                                                                                                                                                                         \n",
    "plot_diff_cop(outdir, dirnames, chrm, start, end, filename='MET10.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nucleosome clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# classify by shift type\n",
    "nuc_df = nuc_diff_map.cluster_nuc_shifts(nuc_df)\n",
    "# cluster by nucleosome posterior\n",
    "nuc_df = nuc_diff_map.cluster_nucs_occ_pdyad(nuc_df, k=4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* **shift_type:** no_shift, depleted, linear_shift, nonlinear_shift\n",
    "* **occ_cluster:** cluster number "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>dyadA</th>\n",
       "      <th>dyadB</th>\n",
       "      <th>occupancyA</th>\n",
       "      <th>occupancyB</th>\n",
       "      <th>pdyadA</th>\n",
       "      <th>pdyadB</th>\n",
       "      <th>shift_AB</th>\n",
       "      <th>dyadC</th>\n",
       "      <th>occupancyC</th>\n",
       "      <th>...</th>\n",
       "      <th>occupancyE</th>\n",
       "      <th>shift_DE</th>\n",
       "      <th>pdyadE</th>\n",
       "      <th>name</th>\n",
       "      <th>shiftAB</th>\n",
       "      <th>shiftBC</th>\n",
       "      <th>shiftCD</th>\n",
       "      <th>shiftDE</th>\n",
       "      <th>shift_type</th>\n",
       "      <th>occ_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>229.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>0.999942</td>\n",
       "      <td>0.984759</td>\n",
       "      <td>0.315144</td>\n",
       "      <td>0.182249</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>0.999783</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999758</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.251794</td>\n",
       "      <td>nuc_0</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>no_shift</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>389.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.999969</td>\n",
       "      <td>0.996902</td>\n",
       "      <td>0.270605</td>\n",
       "      <td>0.136205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.997403</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>0.275946</td>\n",
       "      <td>nuc_1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-19.0</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>no_shift</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>866.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>0.999912</td>\n",
       "      <td>0.427427</td>\n",
       "      <td>0.518984</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999965</td>\n",
       "      <td>...</td>\n",
       "      <td>0.999345</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.277971</td>\n",
       "      <td>nuc_2</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>no_shift</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.889191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuc_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>depleted</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1582.0</td>\n",
       "      <td>1588.0</td>\n",
       "      <td>0.999886</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.310744</td>\n",
       "      <td>0.247687</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1580.0</td>\n",
       "      <td>0.998400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.997392</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.153465</td>\n",
       "      <td>nuc_4</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>no_shift</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   dyadA   dyadB  occupancyA  occupancyB    pdyadA    pdyadB  shift_AB  \\\n",
       "0  chrI   229.0   216.0    0.999942    0.984759  0.315144  0.182249     -13.0   \n",
       "1  chrI   389.0   389.0    0.999969    0.996902  0.270605  0.136205       0.0   \n",
       "2  chrI   866.0   864.0    0.999993    0.999912  0.427427  0.518984      -2.0   \n",
       "3  chrI  1179.0     NaN    0.889191         NaN  0.272899       NaN       NaN   \n",
       "4  chrI  1582.0  1588.0    0.999886    0.999997  0.310744  0.247687       6.0   \n",
       "\n",
       "    dyadC  occupancyC  ...  occupancyE  shift_DE    pdyadE   name  shiftAB  \\\n",
       "0   230.0    0.999783  ...    0.999758       7.0  0.251794  nuc_0    -13.0   \n",
       "1   404.0    0.997403  ...    0.999992      -4.0  0.275946  nuc_1      0.0   \n",
       "2   864.0    0.999965  ...    0.999345       1.0  0.277971  nuc_2     -2.0   \n",
       "3     NaN         NaN  ...         NaN       NaN       NaN  nuc_3      NaN   \n",
       "4  1580.0    0.998400  ...    0.997392      -2.0  0.153465  nuc_4      6.0   \n",
       "\n",
       "   shiftBC  shiftCD  shiftDE  shift_type  occ_cluster  \n",
       "0     14.0     -8.0      7.0    no_shift            2  \n",
       "1     15.0    -19.0     -4.0    no_shift            2  \n",
       "2      0.0      0.0      1.0    no_shift            3  \n",
       "3      NaN      NaN      NaN    depleted            2  \n",
       "4     -8.0      2.0     -2.0    no_shift            2  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuc_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotate nucleosomes as +1, -1, promoter, gene body and intergenic nucleosomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We take annotated +1 and -1 nucleosomes reported by Chereji et al., 2018 to annotate our linked nucleosomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "plus_minus_nucs_file = 'plus_minus_nucs.csv'\n",
    "nuc_df = annotate_nucs.ann_nucs(nuc_df, plus_minus_nucs_file, outdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "New columns:\n",
    "\n",
    "* **+1_nuc:**: Gene name if nucleosome is +1 nucleosome for the gene\n",
    "* **-1_nuc:**: Gene name if nucleosome is -1 nucleosome for the gene\n",
    "* **Promoter_nuc:** Gene name if nucleosome lies within the promoter region for given gene\n",
    "* **ORF_transcript_nuc:** Gene name if nucleosome lies within the ORF or transcript boundary\n",
    "* **Downstream_nuc:** Gene name if nucleosome lies in downstream region of the gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>dyadA</th>\n",
       "      <th>dyadB</th>\n",
       "      <th>occupancyA</th>\n",
       "      <th>occupancyB</th>\n",
       "      <th>pdyadA</th>\n",
       "      <th>pdyadB</th>\n",
       "      <th>shift_AB</th>\n",
       "      <th>dyadC</th>\n",
       "      <th>occupancyC</th>\n",
       "      <th>...</th>\n",
       "      <th>dyadE</th>\n",
       "      <th>occupancyE</th>\n",
       "      <th>shift_DE</th>\n",
       "      <th>pdyadE</th>\n",
       "      <th>name</th>\n",
       "      <th>+1_nuc</th>\n",
       "      <th>-1_nuc</th>\n",
       "      <th>Promoter_nuc</th>\n",
       "      <th>ORF_transcript_nuc</th>\n",
       "      <th>Downstream_nuc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>229.0</td>\n",
       "      <td>216.0</td>\n",
       "      <td>0.999942</td>\n",
       "      <td>0.984759</td>\n",
       "      <td>0.315144</td>\n",
       "      <td>0.182249</td>\n",
       "      <td>-13.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>0.999783</td>\n",
       "      <td>...</td>\n",
       "      <td>229.0</td>\n",
       "      <td>0.999758</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.251794</td>\n",
       "      <td>nuc_0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>389.0</td>\n",
       "      <td>389.0</td>\n",
       "      <td>0.999969</td>\n",
       "      <td>0.996902</td>\n",
       "      <td>0.270605</td>\n",
       "      <td>0.136205</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.997403</td>\n",
       "      <td>...</td>\n",
       "      <td>381.0</td>\n",
       "      <td>0.999992</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>0.275946</td>\n",
       "      <td>nuc_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>866.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>0.999912</td>\n",
       "      <td>0.427427</td>\n",
       "      <td>0.518984</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>864.0</td>\n",
       "      <td>0.999965</td>\n",
       "      <td>...</td>\n",
       "      <td>865.0</td>\n",
       "      <td>0.999345</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.277971</td>\n",
       "      <td>nuc_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1179.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.889191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.272899</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nuc_3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1582.0</td>\n",
       "      <td>1588.0</td>\n",
       "      <td>0.999886</td>\n",
       "      <td>0.999997</td>\n",
       "      <td>0.310744</td>\n",
       "      <td>0.247687</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1580.0</td>\n",
       "      <td>0.998400</td>\n",
       "      <td>...</td>\n",
       "      <td>1580.0</td>\n",
       "      <td>0.997392</td>\n",
       "      <td>-2.0</td>\n",
       "      <td>0.153465</td>\n",
       "      <td>nuc_4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   dyadA   dyadB  occupancyA  occupancyB    pdyadA    pdyadB  shift_AB  \\\n",
       "0  chrI   229.0   216.0    0.999942    0.984759  0.315144  0.182249     -13.0   \n",
       "1  chrI   389.0   389.0    0.999969    0.996902  0.270605  0.136205       0.0   \n",
       "2  chrI   866.0   864.0    0.999993    0.999912  0.427427  0.518984      -2.0   \n",
       "3  chrI  1179.0     NaN    0.889191         NaN  0.272899       NaN       NaN   \n",
       "4  chrI  1582.0  1588.0    0.999886    0.999997  0.310744  0.247687       6.0   \n",
       "\n",
       "    dyadC  occupancyC  ...   dyadE  occupancyE  shift_DE    pdyadE   name  \\\n",
       "0   230.0    0.999783  ...   229.0    0.999758       7.0  0.251794  nuc_0   \n",
       "1   404.0    0.997403  ...   381.0    0.999992      -4.0  0.275946  nuc_1   \n",
       "2   864.0    0.999965  ...   865.0    0.999345       1.0  0.277971  nuc_2   \n",
       "3     NaN         NaN  ...     NaN         NaN       NaN       NaN  nuc_3   \n",
       "4  1580.0    0.998400  ...  1580.0    0.997392      -2.0  0.153465  nuc_4   \n",
       "\n",
       "   +1_nuc  -1_nuc  Promoter_nuc  ORF_transcript_nuc  Downstream_nuc  \n",
       "0     NaN     NaN           NaN                 NaN             NaN  \n",
       "1     NaN     NaN           NaN                 NaN             NaN  \n",
       "2     NaN     NaN           NaN                 NaN             NaN  \n",
       "3     NaN     NaN           NaN                 NaN         YAL068C  \n",
       "4     NaN     NaN           NaN                 NaN         YAL068C  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nuc_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene directional approach"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Nucleosome features for each gene: +1, -1 and gene body nucleosomes. These features for each gene are used to cluster genes in order to obtain gene clusters exhibiting similar nucleosome dynamics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_nucs = nuc_diff_map.cluster_gene_nucs(nuc_df, outdir, 'plus_minus_nucs.csv', k=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Columns in gene nucleosomes:\n",
    "\n",
    "* **p1_shift_XY:** Shift of +1 nucleosome between data sets X and Y\n",
    "\n",
    "* **m1_shift_XY:** Shift of -1 nucleosome between data sets X and Y\n",
    "\n",
    "* **p1_pdyadX:** +1 nucleosome posterior in data set X\n",
    "\n",
    "* **m1_pdyadX:** -1 nucleosome posterior in data set X\n",
    "\n",
    "* **orf_pdyadX:** Average nucleosome dyad posterior in gene/transcript\n",
    "\n",
    "* **cluster:** Assigned cluster\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1_shift_AB</th>\n",
       "      <th>p1_shift_BC</th>\n",
       "      <th>p1_shift_CD</th>\n",
       "      <th>p1_shift_DE</th>\n",
       "      <th>m1_shift_AB</th>\n",
       "      <th>m1_shift_BC</th>\n",
       "      <th>m1_shift_CD</th>\n",
       "      <th>m1_shift_DE</th>\n",
       "      <th>p1_pdyadA</th>\n",
       "      <th>p1_pdyadB</th>\n",
       "      <th>...</th>\n",
       "      <th>m1_pdyadB</th>\n",
       "      <th>m1_pdyadC</th>\n",
       "      <th>m1_pdyadD</th>\n",
       "      <th>m1_pdyadE</th>\n",
       "      <th>orf_pdyadA</th>\n",
       "      <th>orf_pdyadB</th>\n",
       "      <th>orf_pdyadC</th>\n",
       "      <th>orf_pdyadD</th>\n",
       "      <th>orf_pdyadE</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YAL001C</th>\n",
       "      <td>7.0</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.447447</td>\n",
       "      <td>0.230651</td>\n",
       "      <td>...</td>\n",
       "      <td>0.312737</td>\n",
       "      <td>0.341480</td>\n",
       "      <td>0.261071</td>\n",
       "      <td>0.598655</td>\n",
       "      <td>0.323751</td>\n",
       "      <td>0.327657</td>\n",
       "      <td>0.330314</td>\n",
       "      <td>0.335367</td>\n",
       "      <td>0.324313</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL002W</th>\n",
       "      <td>5.0</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.460655</td>\n",
       "      <td>0.397216</td>\n",
       "      <td>...</td>\n",
       "      <td>0.267551</td>\n",
       "      <td>0.426574</td>\n",
       "      <td>0.294886</td>\n",
       "      <td>0.130260</td>\n",
       "      <td>0.395948</td>\n",
       "      <td>0.304407</td>\n",
       "      <td>0.305774</td>\n",
       "      <td>0.343116</td>\n",
       "      <td>0.329954</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL003W</th>\n",
       "      <td>10.0</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>-71.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-6.0</td>\n",
       "      <td>0.247866</td>\n",
       "      <td>0.430759</td>\n",
       "      <td>...</td>\n",
       "      <td>0.250443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.455934</td>\n",
       "      <td>0.216981</td>\n",
       "      <td>0.135086</td>\n",
       "      <td>0.203688</td>\n",
       "      <td>0.159156</td>\n",
       "      <td>0.232781</td>\n",
       "      <td>0.253109</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL005C</th>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-10.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>0.405285</td>\n",
       "      <td>0.250443</td>\n",
       "      <td>...</td>\n",
       "      <td>0.430759</td>\n",
       "      <td>0.276569</td>\n",
       "      <td>0.369917</td>\n",
       "      <td>0.337704</td>\n",
       "      <td>0.370470</td>\n",
       "      <td>0.294824</td>\n",
       "      <td>0.137953</td>\n",
       "      <td>0.289722</td>\n",
       "      <td>0.271661</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YAL007C</th>\n",
       "      <td>-7.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-14.0</td>\n",
       "      <td>0.281245</td>\n",
       "      <td>0.242525</td>\n",
       "      <td>...</td>\n",
       "      <td>0.352263</td>\n",
       "      <td>0.238988</td>\n",
       "      <td>0.218626</td>\n",
       "      <td>0.182860</td>\n",
       "      <td>0.299204</td>\n",
       "      <td>0.349682</td>\n",
       "      <td>0.345075</td>\n",
       "      <td>0.413458</td>\n",
       "      <td>0.351676</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         p1_shift_AB  p1_shift_BC  p1_shift_CD  p1_shift_DE  m1_shift_AB  \\\n",
       "gene                                                                       \n",
       "YAL001C          7.0         -0.0         -1.0          3.0         10.0   \n",
       "YAL002W          5.0         -6.0          5.0         10.0         21.0   \n",
       "YAL003W         10.0        -10.0        -71.0          0.0         -8.0   \n",
       "YAL005C          8.0          NaN          NaN          6.0        -10.0   \n",
       "YAL007C         -7.0          7.0          8.0         -0.0         -0.0   \n",
       "\n",
       "         m1_shift_BC  m1_shift_CD  m1_shift_DE  p1_pdyadA  p1_pdyadB  ...  \\\n",
       "gene                                                                  ...   \n",
       "YAL001C         -0.0         -9.0          9.0   0.447447   0.230651  ...   \n",
       "YAL002W          0.0          4.0          6.0   0.460655   0.397216  ...   \n",
       "YAL003W          NaN          NaN         -6.0   0.247866   0.430759  ...   \n",
       "YAL005C         10.0         71.0         -0.0   0.405285   0.250443  ...   \n",
       "YAL007C         -1.0          2.0        -14.0   0.281245   0.242525  ...   \n",
       "\n",
       "         m1_pdyadB  m1_pdyadC  m1_pdyadD  m1_pdyadE  orf_pdyadA  orf_pdyadB  \\\n",
       "gene                                                                          \n",
       "YAL001C   0.312737   0.341480   0.261071   0.598655    0.323751    0.327657   \n",
       "YAL002W   0.267551   0.426574   0.294886   0.130260    0.395948    0.304407   \n",
       "YAL003W   0.250443        NaN   0.455934   0.216981    0.135086    0.203688   \n",
       "YAL005C   0.430759   0.276569   0.369917   0.337704    0.370470    0.294824   \n",
       "YAL007C   0.352263   0.238988   0.218626   0.182860    0.299204    0.349682   \n",
       "\n",
       "         orf_pdyadC  orf_pdyadD  orf_pdyadE  cluster  \n",
       "gene                                                  \n",
       "YAL001C    0.330314    0.335367    0.324313        4  \n",
       "YAL002W    0.305774    0.343116    0.329954        1  \n",
       "YAL003W    0.159156    0.232781    0.253109        6  \n",
       "YAL005C    0.137953    0.289722    0.271661        6  \n",
       "YAL007C    0.345075    0.413458    0.351676        1  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_nucs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TFs that are differentially enriched in the promoter region of the gene clusters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_set = get_diff_tfs.get_tf_clusters(gene_nucs, dirnames, outdir, 'plus_minus_nucs.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Table with the number of columns equal to the number of clusters identified using ```cluster_gene_nucs```. Each row corresponds to one TF. Each cell represents the fraction of genes and in the given cluster that have differential activity of the TF of that row."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ABF1</th>\n",
       "      <td>11.306533</td>\n",
       "      <td>7.050093</td>\n",
       "      <td>9.423233</td>\n",
       "      <td>12.380952</td>\n",
       "      <td>9.917355</td>\n",
       "      <td>9.432146</td>\n",
       "      <td>11.016949</td>\n",
       "      <td>8.035714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASG1</th>\n",
       "      <td>19.849246</td>\n",
       "      <td>7.513915</td>\n",
       "      <td>9.991877</td>\n",
       "      <td>19.761905</td>\n",
       "      <td>9.917355</td>\n",
       "      <td>10.009625</td>\n",
       "      <td>19.491525</td>\n",
       "      <td>8.928571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DAL82</th>\n",
       "      <td>8.542714</td>\n",
       "      <td>4.359926</td>\n",
       "      <td>6.742486</td>\n",
       "      <td>11.190476</td>\n",
       "      <td>4.752066</td>\n",
       "      <td>5.101059</td>\n",
       "      <td>16.949153</td>\n",
       "      <td>6.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAL4</th>\n",
       "      <td>13.316583</td>\n",
       "      <td>4.638219</td>\n",
       "      <td>8.204712</td>\n",
       "      <td>15.238095</td>\n",
       "      <td>6.714876</td>\n",
       "      <td>6.833494</td>\n",
       "      <td>21.186441</td>\n",
       "      <td>6.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCR1</th>\n",
       "      <td>16.834171</td>\n",
       "      <td>6.957328</td>\n",
       "      <td>9.748172</td>\n",
       "      <td>15.476190</td>\n",
       "      <td>6.818182</td>\n",
       "      <td>6.544755</td>\n",
       "      <td>24.576271</td>\n",
       "      <td>5.357143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               1         2         3          4         5          6  \\\n",
       "ABF1   11.306533  7.050093  9.423233  12.380952  9.917355   9.432146   \n",
       "ASG1   19.849246  7.513915  9.991877  19.761905  9.917355  10.009625   \n",
       "DAL82   8.542714  4.359926  6.742486  11.190476  4.752066   5.101059   \n",
       "GAL4   13.316583  4.638219  8.204712  15.238095  6.714876   6.833494   \n",
       "GCR1   16.834171  6.957328  9.748172  15.476190  6.818182   6.544755   \n",
       "\n",
       "               7         8  \n",
       "ABF1   11.016949  8.035714  \n",
       "ASG1   19.491525  8.928571  \n",
       "DAL82  16.949153  6.250000  \n",
       "GAL4   21.186441  6.250000  \n",
       "GCR1   24.576271  5.357143  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tf_set.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
